src/protocol/aggregation/mod.rs

```diff
@@ -26,110 +27,124 @@ use crate::{
-        .collect::<Vec<_>>();
-
-    secure_aggregation(validator, binary_shared_values).await
+
+    // TODO(taikiy):
+    // 1. slice the buckets into N streams and send them to the aggregation protocol
+    // 2. collect the results and return them
+
+    let output = aggregate_values(validator.context(), gf2_value_bits).await?;
+
+    validator.validate(vec![output]).await
 }
 
 /// Performs a set of aggregation protocols on binary shared values.
+/// This protocol assumes that the input rows are from the same match key and
+/// for the same breakdown key, so that we don't have to worry about computing
+/// the helper_bits or capping.
 ///
 /// # Errors
 /// propagates errors from multiplications
-#[tracing::instrument(name = "aggregate", skip_all)]
-pub async fn secure_aggregation<V, VB, C, SB, F>(
-    _binary_validator: VB,
-    _binary_shared_values: Vec<BinarySharedAggregateInputs<SB>>,
-) -> Result<Vec<Replicated<F>>, Error>
+#[tracing::instrument(name = "simple_aggregate_values", skip_all)]
+pub async fn aggregate_values<F, C, S>(
+    ctx: C,
+    contribution_value_bits_gf2: Vec<BitDecomposed<Replicated<Gf2>>>,
+) -> Result<S, Error>
 where
-    VB: Validator<C, Gf2>,
-    C: UpgradableContext<Validator<F> = V>,
-    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = SB> + Context,
-    SB: LinearSecretSharing<Gf2>
-        + BasicProtocols<C::UpgradedContext<Gf2>, Gf2>
-        + DowngradeMalicious<Target = Replicated<Gf2>>
-        + 'static,
-    F: PrimeField + ExtendableField,
+    F: PrimeField,
+    C: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + BasicProtocols<C, F> + Serializable + 'static,
 {
-    // TODO(taikiy):
-    // 0. This protocol assumes that the input rows are from the same match key
-    //    so that we don't have to worry about computing the helper_bits. There
-    //    will be some pre-processing to do before this protocol is called.
-    // 1. validate the binary shares
-    // 2. mod convert
-    // 3. apply capping
-    // 4. aggregate per breakdown key
-    // 5. validate the result before returning
-
-    todo!()
+    let record_count = contribution_value_bits_gf2.len();
+    let bits = contribution_value_bits_gf2[0].len();
+
+    // mod-convert for later validation
+    let convert_ctx = ctx
+        .narrow(&Step::ConvertValueBits)
+        .set_total_records(record_count);
+    let converted_contribution_values = convert_bits(
+        convert_ctx,
+        stream_iter(contribution_value_bits_gf2),
+        0..u32::try_from(bits).unwrap(),
+    );
+
+    //TODO: retrieve the previous aggregate value from the storage
+
+    let aggregate = converted_contribution_values
+        .try_fold(S::ZERO, |mut acc, row| async move {
+            acc += &row.as_decimal();
+            Ok(acc)
+        })
+        .await?;
+
+    //TODO:
+    // * add the new aggregate value to the previous one
+    // * add DP noise relative to cap and max # of reports
```

Comment:
No. The DP noise is not related in any way to the # of reports. It's only related to the cap.